ipam/provider/aws.go

```diff
@@ -0,0 +1,144 @@
+package provider
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/giantswarm/e2e-harness/pkg/framework"
+	awsclient "github.com/giantswarm/e2eclients/aws"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+)
+
+// apiextensionsAWSConfigE2EChartValues is modified version of
+// e2etemplates.ApiextensionsAWSConfigE2EChartValues for IPAM test scenario.
+// Major difference is lack of subnet fields and need to replace clusterName
+// format string value. ClusterName is not based on env variable because there
+// would be too high risk for unintentional disruption of other tests if
+// ${CLUSTER_NAME} would be dynamically changed on the fly.
+const apiextensionsAWSConfigE2EChartValues = `commonDomain: ${COMMON_DOMAIN}
+clusterName: %s
+clusterVersion: v_0_1_0
+sshPublicKey: ${IDRSA_PUB}
+versionBundleVersion: ${VERSION_BUNDLE_VERSION}
+aws:
+  region: ${AWS_REGION}
+  apiHostedZone: ${AWS_API_HOSTED_ZONE_GUEST}
+  ingressHostedZone: ${AWS_INGRESS_HOSTED_ZONE_GUEST}
+  routeTable0: ${AWS_ROUTE_TABLE_0}
+  routeTable1: ${AWS_ROUTE_TABLE_1}
+  vpcPeerId: ${AWS_VPC_PEER_ID}
+`
+
+type AWSConfig struct {
+	AWSClient     *awsclient.Client
+	HostFramework *framework.Host
+	Logger        micrologger.Logger
+}
+
+type AWS struct {
+	awsClient     *awsclient.Client
+	hostFramework *framework.Host
+	logger        micrologger.Logger
+}
+
+func NewAWS(config AWSConfig) (*AWS, error) {
+	if config.AWSClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.AWSClient must not be empty", config)
+	}
+	if config.HostFramework == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.HostFramework must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	a := &AWS{
+		awsClient:     config.AWSClient,
+		hostFramework: config.HostFramework,
+		logger:        config.Logger,
+	}
+
+	return a, nil
+}
+
+func (aws *AWS) RequestGuestClusterCreation(clusterName string) error {
+	if clusterName == "" {
+		return microerror.Maskf(invalidConfigError, "clusterName must not be empty")
+	}
+
+	deploymentName := awsConfigDeploymentName(clusterName)
+
+	o := func() error {
+		// NOTE we ignore errors here because we cannot get really useful error
+		// handling done. This here should anyway only be a quick fix until we use
+		// the helm client lib. Then error handling will be better.
+		framework.HelmCmd(fmt.Sprintf("delete --purge %s", deploymentName))
```

Comment:
Same here as with the env vars. Lets stop doing the shell out bullshit. We can just use a helm client. We have the library. We have the frameworks. That is all there and set up. And then we also have some proper error handling. 

I also think that the resource creation should be abstracted somehow. We do that the same way like in all e2e tests so there is some common ground to cover.

Reply:
Can you elaborate, which helm client and are there already examples how to use it? This piece of code was from `aws-operator's` present e2e-test setup: https://github.com/giantswarm/aws-operator/blob/master/integration/setup/setup.go#L172-L175

Regarding abstraction here, working on e2e test framework would be too large task within the scope of my current story although I do agree with you regarding interface here.

Reply:
The guest and hist frameworks already expose a k8s and a g8s client. They have a rest config. The frameworks should then also just expose a helm client which is also set up with the rest. The necessary magic is already done in https://github.com/giantswarm/guestcluster.

Reply:
Using the helm client to delete a chart is done in `chart-operator`. See https://github.com/giantswarm/chart-operator/blob/9d0955af54e11dbd5c6661acf8ac4ecc3a92b4b1/service/controller/v3/resource/chart/delete.go.

Reply:
I totally understand you and agree here, but this change is pretty big and should be done as a separate project in SIG Operator or SIG Releng in order to cover relevant fields in `e2e-harness`. For the scope of this PR, in order to get `aws-operator` IPAM implementation eventually ready, I'd rather leave this as-is.

Reply:
Can we have a TODO issue for these things so we track the problem and its addressed progress with all the places where we have to fix it.